File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuffer sb = new StringBuffer();

        String[] testCase = br.readLine().split(" ");
        int k = Integer.parseInt(testCase[0]);
        int n = Integer.parseInt(testCase[1]);
        int[] lans = new int[k];
        long max = 0;
        long min = 1;

        for (int i = 0; i < k; i++) {
            lans[i] = Integer.parseInt(br.readLine());

            if (max < lans[i]) {
                max = lans[i];
            }
        }

        while (min <= max) {
            long count = 0;
            long mid = (max + min) / 2;
            for (int i = 0; i < k; i++) {
                count += lans[i] / mid;
            }

            if (n > count) {
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }
        sb.append(max);

        bw.write(sb.toString());
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.c
#include <stdio.h>

int main() {
    int n, k;
    
    scanf("%d %d", &n, &k);
    
    int wires[n];
    long start = 1, end = 1;
    
    for (int i = 0; i < n; i++) {
        scanf("%ld", &wires[i]);
        if (wires[i] > end) end = wires[i];
    }

    while(start <= end) { // 이분 탐색으로 최대 길이를 구함
        long mid = (start+end)/2;
        
        long count = 0;
        for (int i = 0; i < n; i++) {
            count += wires[i]/mid;
        }

        if (count < k) end = mid-1; // 개수가 부족하면 더 작은 길이로 잘라야 하니까
        else start = mid + 1;
    }

    printf("%ld",  end);
}